src/java/com/puppetlabs/puppetserver/pool/JRubyPool.java

```diff
@@ -1,234 +1,371 @@
 package com.puppetlabs.puppetserver.pool;
 
+import java.util.LinkedList;
 import java.util.Set;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.CopyOnWriteArraySet;
-import java.util.concurrent.LinkedBlockingDeque;
-import java.util.concurrent.locks.ReentrantReadWriteLock;
-import com.puppetlabs.puppetserver.pool.LockablePool;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
 
 /**
- * A data structure to be used as a Pool, encapsulating a LinkedBlockingDeque
- * and a ReentrantReadWriteLock. Implements the <tt>LockablePool</tt>
- * interface such that when a <tt>borrowItem</tt> is performed an instance is
- * taken from the deque and a read lock is acquired, when a
- * <tt>releaseItem</tt> is performed an instance is put back onto the deque and
- * a read lock is released, and <tt>lock</tt> locks a write lock.
+ * An implementation of LockablePool for managing a pool of JRubyPuppet
+ * instances.
  *
- * @param <E> the type of element that can be added to the LinkedBlockingDeque.
+ * @param <E> the type of element that can be added to the pool.
  */
 public final class JRubyPool<E> implements LockablePool<E> {
-    private final LinkedBlockingDeque<E> liveQueue;
-    private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock(true);
-    private final Set<E> registeredElements = new CopyOnWriteArraySet<>();
+    // The `LockingPool` contract requires some synchronization behaviors that
+    // are not natively present in any of the JDK deque implementations -
+    // specifically to allow one calling thread to call lock() to supercede
+    // and hold off any pending pool borrowers until unlock() is called.
+    // This class implementation fulfills the contract by managing the
+    // synchronization constructs directly rather than deferring to an
+    // underlying JDK data structure to manage concurrent access.
+    //
+    // This implementation is modeled somewhat off of what the
+    // `LinkedBlockingDeque` class in the OpenJDK does to manage
+    // concurrency.  It uses a single `ReentrantLock` to provide mutual
+    // exclusion around offer and take requests, with condition variables
+    // used to park and later reawaken requests as needed, e.g., when pool
+    // items are unavailable for borrowing or when the write lock is
+    // unavailable.
+    //
+    // See http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/util/concurrent/LinkedBlockingDeque.java#l157
+    //
+    // Because access to the underlying deque is synchronized within
+    // this class, the pool is backed by a non-synchronized JDK `LinkedList`.
 
-    public JRubyPool(int size) {
-        liveQueue = new LinkedBlockingDeque<>(size);
-    }
+    // Underlying queue which holds the elements that clients can borrow.
+    private final LinkedList<E> liveQueue;
```

Comment:
> 1) Can this type be just List instead of LinkedList or does the implementation depend on some methods on LinkedList?

The specific methods the implementation depends upon come from the `Deque<E>` interface.  I'll change the definition to use that instead to make it a little more generic.

> 2) If the answer to that is "yes", why LinkedList instead of the more commonly-used ArrayList?

The properties of being able to insert an element at the front and back of the queue are what I believe had led us to migrating from having this structure be a queue to a deque - currently as a `LinkedBlockingDeque`.  Migrating the concurrency protection up into the `JRubyPool` class, I wanted to go with a non-synchronized backing class as a replacement, to avoid the performance overhead of double-synchronization.  I had been thinking a `LinkedList` would be better than an `ArrayList` because the operation of inserting an item at the beginning of the collection, which is what we most commonly do when returning a borrowed element to the pool, could be significantly slower with the `ArrayList` since it would involve shifting all of the existing elements down to accommodate the new one.

I hadn't, however, considered `ArrayDeque` instead, added in Java 6.  `ArrayDeque` satisfies the `Deque<E>` interface and, according to a few references I've seen, seems to be superior to a `LinkedList` from a performance perspective for the general operations we would use.  From https://docs.oracle.com/javase/tutorial/collections/implementations/deque.html:

> In terms of efficiency, ArrayDeque is more efficient than the LinkedList for add and remove operation at both ends. The best operation in a LinkedList implementation is removing the current element during the iteration. LinkedList implementations are not ideal structures to iterate.  The LinkedList implementation consumes more memory than the ArrayDeque implementation.

`ArrayDeque` might be a worse choice than `LinkedList` if we were having to resize / copy content frequently but where our queue is fixed size (at least today), `ArrayDeque` might be a better choice.  Maybe if we need to do dynamic resizing later and the need for / cost of resizing the `ArrayDeque` frequently were prohibitively expensive, it would be best to go with a `LinkedList`.  But we could wait until later if that turned out to be an issue.

I'll try out `ArrayDeque` instead and go with that instead if it looks workable.  What do you think?